Sources/PackageCollectionsSigning/Utilities/Base64URL.swift

```diff
@@ -0,0 +1,88 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Vapor open source project
+//
+// Copyright (c) 2017-2020 Vapor project authors
+// Licensed under MIT
+//
+// See LICENSE for license information
+//
+// SPDX-License-Identifier: MIT
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+// Source: https://github.com/vapor/jwt-kit/blob/master/Sources/JWTKit/Utilities/Base64URL.swift
+
+extension DataProtocol {
+    func base64URLDecodedBytes() -> [UInt8] {
+        return Data(base64Encoded: Data(self.copyBytes()).base64URLUnescaped())?.copyBytes() ?? []
```

Comment:
By my count this one line of code allocates 5 times and copies 4 times:

1. `self.copyBytes()` will allocate
2. `Data()` then allocates to copy those bytes
3. `.base64URLUnescaped()` then allocates again
4. `Data(base64Encoded:)` then allocates again
5. Then the outer `copyBytes` allocates once more.

I have mostly not bothered making performance notes in this code review but I think 5 allocations and 4 copies in one line seems like we should intervene to address this.

The first `copyBytes` is unnecessary: `DataProtocol` is always a `Collection`, so we can just construct a `Data` from it directly.  Secondly, `base64URLUnescaped` makes a copy in order to mutate it, but we don't need to do that: we uniquely own the `Data` here (we just copied it!) so we can do the modification in-place.

Finally, is it really necessary that this returns a `[UInt8]`? Could we return a `Data` here to avoid this copy? If we could then we'd be able to elide three allocations and two copies.

Reply:
Cleaned this up in https://github.com/apple/swift-package-manager/pull/3242/commits/667c95473df57c837a561043821750a42cf28691